File: aznavrail-web/src/utils/RelocItemHandler.js

export const RelocItemHandler = {
  findCluster: (items, itemId, knownIndex = -1) => {
    let index = knownIndex;
    if (index === -1 || index >= items.length || items[index].id !== itemId) {
        index = items.findIndex(it => it.id === itemId);
    }
    if (index === -1) return null;

    const item = items[index];
    if (!item.isRelocItem || !item.hostId) return null;

    let start = index;
    while (start > 0) {
      const prev = items[start - 1];
      if (prev.isRelocItem && prev.hostId === item.hostId) {
        start--;
      } else {
        break;
      }
    }

    let end = index;
    while (end < items.length - 1) {
      const next = items[end + 1];
      if (next.isRelocItem && next.hostId === item.hostId) {
        end++;
      } else {
        break;
      }
    }

    return { start, end };
  },

  updateOrder: (items, draggedId, targetIndex, knownCurrentIndex = -1) => {
    const currentIndex = knownCurrentIndex !== -1 ? knownCurrentIndex : items.findIndex(it => it.id === draggedId);
    if (currentIndex === -1 || currentIndex === targetIndex) return items;

    // Check cluster
    const cluster = RelocItemHandler.findCluster(items, draggedId, currentIndex);
    if (!cluster || targetIndex < cluster.start || targetIndex > cluster.end) return items;

    const newItems = [...items];
    const [item] = newItems.splice(currentIndex, 1);
    newItems.splice(targetIndex, 0, item);
    return newItems;
  }
};

File: aznavrail/src/main/java/com/hereliesaz/aznavrail/internal/RelocItemHandler.kt
package com.hereliesaz.aznavrail.internal

import androidx.compose.ui.geometry.Rect
import com.hereliesaz.aznavrail.model.AzNavItem

object RelocItemHandler {
    val itemBoundsCache = mutableMapOf<String, Rect>()

    fun findCluster(items: List<AzNavItem>, itemId: String): IntRange? {
        val index = items.indexOfFirst { it.id == itemId }
        if (index == -1) return null

        val item = items[index]
        if (!item.isRelocItem || item.hostId == null) return null

        var start = index
        while (start > 0 && items[start - 1].isRelocItem && items[start - 1].hostId == item.hostId) {
            start--
        }

        var end = index
        while (end < items.size - 1 && items[end + 1].isRelocItem && items[end + 1].hostId == item.hostId) {
            end++
        }

        return start..end
    }

    fun updateOrder(items: MutableList<AzNavItem>, draggedId: String, targetIndex: Int) {
        val currentIndex = items.indexOfFirst { it.id == draggedId }
        if (currentIndex == -1 || currentIndex == targetIndex) return

        val cluster = findCluster(items, draggedId) ?: return
        if (targetIndex !in cluster) return

        val item = items.removeAt(currentIndex)
        items.add(targetIndex, item)
    }

    fun calculateTargetIndex(
        items: List<AzNavItem>,
        draggedItemId: String,
        currentDragOffset: Float,
        itemBounds: Map<String, Rect>,
        isVertical: Boolean
    ): Int? {
        val currentIndex = items.indexOfFirst { it.id == draggedItemId }
        if (currentIndex == -1) return null

        val draggedBounds = itemBounds[draggedItemId] ?: return null
        val cluster = findCluster(items, draggedItemId) ?: return null

        val draggedCenter = if (isVertical) {
            draggedBounds.center.y + currentDragOffset
        } else {
            draggedBounds.center.x + currentDragOffset
        }

        var bestTarget = currentIndex
        var minDistance = Float.MAX_VALUE

        for (i in cluster) {
            val item = items[i]
            val bounds = itemBounds[item.id]
            if (bounds != null) {
                val itemCenter = if (isVertical) bounds.center.y else bounds.center.x
                val distance = kotlin.math.abs(draggedCenter - itemCenter)
                if (distance < minDistance) {
                    minDistance = distance
                    bestTarget = i
                }
            }
        }
        return bestTarget
    }
}
